feg/gateway/services/envoy_controller/control_plane/control_plane.go

```diff
@@ -0,0 +1,364 @@
+	orig_src "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/listener/original_src/v3"
+
+	"github.com/envoyproxy/go-control-plane/pkg/wellknown"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/wrappers"
+	"magma/feg/cloud/go/protos"
+)
+
+type EnvoyController interface {
+	UpdateSnapshot([]*protos.AddUEHeaderEnrichmentRequest)
+}
+
+type ControllerClient struct {
+	version int32
+	config  cache.SnapshotCache
+}
+
+const (
+	XdsCluster                  = "xds_cluster"
+	Ads                         = "ads"
+	Xds                         = "xds"
+	Rest                        = "rest"
+	debug                       = true
+	port                        = 18000
+	gatewayPort                 = 18001
+	mode                        = Ads
+	any_addr                    = "0.0.0.0"
+	maxConcurrentStreams        = 16
+	initialStreamWindowSize     = 65536   // 64Kib
+	initialConnectionWindowSize = 1048576 // 1 MiB
+	connectTimeout              = 6 * time.Second
+	idleTimeout                 = 3600 * time.Second
+	httpPort                    = 80
+	clusterName                 = "cluster1"
+	targetPrefix                = "/"
+)
+
+type logger struct{}
+
+func (logger logger) Infof(format string, args ...interface{}) {
+	glog.Infof(format, args...)
+}
+func (logger logger) Errorf(format string, args ...interface{}) {
+	glog.Errorf(format, args...)
+}
+func (cb *callbacks) Report() {
+	cb.mu.Lock()
+	defer cb.mu.Unlock()
+	glog.Infof("cb.Report() fetches %d,  callbacks %d", cb.fetches, cb.requests)
+}
+func (cb *callbacks) OnStreamOpen(ctx context.Context, id int64, typ string) error {
+	glog.Infof("OnStreamOpen %d open for %s", id, typ)
+	return nil
+}
+func (cb *callbacks) OnStreamClosed(id int64) {
+	glog.Infof("OnStreamClosed %d closed", id)
+}
+func (cb *callbacks) OnStreamRequest(int64, *v2.DiscoveryRequest) error {
+	glog.Infof("OnStreamRequest")
+	cb.mu.Lock()
+	defer cb.mu.Unlock()
+	cb.requests++
+	if cb.signal != nil {
+		close(cb.signal)
+		cb.signal = nil
+	}
+	return nil
+}
+func (cb *callbacks) OnStreamResponse(int64, *v2.DiscoveryRequest, *v2.DiscoveryResponse) {
+	glog.Infof("OnStreamResponse...")
+	cb.Report()
+}
+func (cb *callbacks) OnFetchRequest(ctx context.Context, req *v2.DiscoveryRequest) error {
+	glog.Infof("OnFetchRequest...")
+	cb.mu.Lock()
+	defer cb.mu.Unlock()
+	cb.fetches++
+	if cb.signal != nil {
+		close(cb.signal)
+		cb.signal = nil
+	}
+	return nil
+}
+func (cb *callbacks) OnFetchResponse(*v2.DiscoveryRequest, *v2.DiscoveryResponse) {}
+
+type callbacks struct {
+	signal   chan struct{}
+	fetches  int
+	requests int
+	mu       sync.Mutex
+}
+
+// Hasher returns node ID as an ID
+type Hasher struct {
+}
+
+// ID function
+func (h Hasher) ID(node *core.Node) string {
+	if node == nil {
+		return "unknown"
+	}
+	return node.Id
+}
+
+const grpcMaxConcurrentStreams = 1000000
+
+// RunManagementServer starts an xDS server at the given port.
```

Comment:
I am probably missing something, but is `RunManagementServer` the grpcServer itself? I am asking because I see 2 GRPC server (the one created here and another being created by `NewServiceWithOptions`)

Reply:
yeah I think we should keep these grpc services separate. This one is envoy managment and the other is the actual envoy controller magma service grpc server